File: app/src/main/java/com/jcr/sharedtasks/ui/list/TasksListFragment.java
package com.jcr.sharedtasks.ui.list;

import android.app.Activity;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.databinding.DataBindingComponent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;

import com.jcr.sharedtasks.R;
import com.jcr.sharedtasks.binding.FragmentDataBindingComponent;
import com.jcr.sharedtasks.databinding.TasksListFragmentBinding;
import com.jcr.sharedtasks.di.Injectable;
import com.jcr.sharedtasks.ui.common.NavigationController;
import com.jcr.sharedtasks.util.AutoClearedValue;
import com.jcr.sharedtasks.widget.TasksListWidgetService;

import javax.inject.Inject;

import static com.jcr.sharedtasks.BR.projectTasksList;

public class TasksListFragment extends Fragment implements Injectable {

    private static final String PROJECT_UUID_KEY = "projectUUID";

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    @Inject
    NavigationController navigationController;

    DataBindingComponent dataBindingComponent = new FragmentDataBindingComponent(this);

    AutoClearedValue<TasksListFragmentBinding> binding;

    AutoClearedValue<TasksListAdapter> adapter;

    private TasksListViewModel tasksListViewModel;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        TasksListFragmentBinding dataBinding = DataBindingUtil
                .inflate(inflater, R.layout.tasks_list_fragment, container, false,
                        dataBindingComponent);
        binding = new AutoClearedValue<>(this, dataBinding);

        return dataBinding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        tasksListViewModel = ViewModelProviders.of(this, viewModelFactory).get(TasksListViewModel.class);
        TasksListAdapter tasksListAdapter = new TasksListAdapter(dataBindingComponent,
                task -> navigationController.navigateToTaskDetail(task.getTaskSID()),
                tasksListViewModel::updateTaskStatus,
                tasksListViewModel::updateTaskAssignee);

        Bundle args = getArguments();
        if (args.containsKey(PROJECT_UUID_KEY)) {
            tasksListViewModel.setProjectUUID(args.getString(PROJECT_UUID_KEY));
        }

        binding.get().tasksListRv.setAdapter(tasksListAdapter);
        adapter = new AutoClearedValue<>(this, tasksListAdapter);

        fillViews();
        TasksListWidgetService.startActionUpdateIngredientsList(getContext());
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_task_list, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.add_people:
                sendInvite();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onResume() {
        super.onResume();

        hideKeyboard();
    }

    private void hideKeyboard() {
        Activity activity = getActivity();
        InputMethodManager imm = (InputMethodManager) activity.getSystemService(Activity.INPUT_METHOD_SERVICE);
        View view = activity.getCurrentFocus();
        if (view == null) {
            view = new View(activity);
        }
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }

    private void sendInvite() {
        Intent sendIntent = new Intent();
        String msg = getString(R.string.invite_content);
        sendIntent.setAction(Intent.ACTION_SEND);
        sendIntent.putExtra(Intent.EXTRA_TEXT, msg + tasksListViewModel.getDeepLinkOfCurrentProject());
        sendIntent.setType("text/plain");
        startActivity(sendIntent);
    }

    private void fillViews() {
        tasksListViewModel.getProjectReference().observe(this,
                projectReference -> {
                    if (projectReference != null) {
                        setupActionBar(projectReference.getProjectName());
                    }
                });

        binding.get().addTask.setOnClickListener(v -> onNewTaskClick());
        initRecyclerView();
    }

    private void setupActionBar(String projectName) {
        ActionBar actionBar = ((AppCompatActivity)getActivity()).getSupportActionBar();
        if (actionBar == null) {
            return;
        }
        actionBar.setTitle(projectName);
    }

    private void initRecyclerView() {
        tasksListViewModel.getTasks().observe(this, tasks -> {
            if (tasks != null) {
                binding.get().setVariable(projectTasksList, tasks);
                adapter.get().replace(tasks);
                binding.get().executePendingBindings();
            }
        });
    }

    public void onNewTaskClick() {
        navigationController.navigateToTaskDetail(null);
    }

    public static TasksListFragment create(String projectUUID) {
        TasksListFragment tasksListFragment = new TasksListFragment();
        Bundle args = new Bundle();
        args.putString(PROJECT_UUID_KEY, projectUUID);
        tasksListFragment.setArguments(args);
        return tasksListFragment;
    }

}

File: app/src/main/java/com/jcr/sharedtasks/ui/list/TasksListFragment.kt
package com.jcr.sharedtasks.ui.list

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.view.inputmethod.InputMethodManager
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingComponent
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import com.jcr.sharedtasks.AppExecutors
import com.jcr.sharedtasks.R
import com.jcr.sharedtasks.binding.FragmentDataBindingComponent
import com.jcr.sharedtasks.databinding.TasksListFragmentBinding
import com.jcr.sharedtasks.di.Injectable
import com.jcr.sharedtasks.ui.common.NavigationController
import com.jcr.sharedtasks.util.autoCleared
import com.jcr.sharedtasks.util.collectOnStarted
import com.jcr.sharedtasks.widget.TasksListWidgetService
import javax.inject.Inject

open class TasksListFragment : Fragment(), Injectable {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    @Inject
    lateinit var navigationController: NavigationController

    @Inject
    lateinit var appExecutors: AppExecutors

    var dataBindingComponent: DataBindingComponent = FragmentDataBindingComponent(this)

    var binding by autoCleared<TasksListFragmentBinding>()

    var adapter by autoCleared<TasksListAdapter>()

    lateinit var tasksListViewModel: TasksListViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val dataBinding = DataBindingUtil
                .inflate<TasksListFragmentBinding>(inflater, R.layout.tasks_list_fragment, container, false,
                        dataBindingComponent)

        binding = dataBinding

        return dataBinding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        tasksListViewModel = ViewModelProviders.of(this, viewModelFactory).get(TasksListViewModel::class.java)
        val tasksListAdapter = TasksListAdapter(
                dataBindingComponent,
                appExecutors,
                { task ->
                    navigationController.navigateToTaskDetail(task.taskSID)
                },
                { task ->
                    tasksListViewModel.updateTaskStatus(task)
                },
                { task ->
                    tasksListViewModel.updateTaskAssignee(task)
                })

        val args = arguments
        if (args!!.containsKey(PROJECT_UUID_KEY)) {
            tasksListViewModel.setProjectUUID(args.getString(PROJECT_UUID_KEY)!!)
        }

        binding.tasksListRv.adapter = tasksListAdapter
        adapter = tasksListAdapter

        fillViews()
        initWidget()
    }

    open fun initWidget() {
//        TasksListWidgetService.startActionUpdateIngredientsList(context)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_task_list, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.add_people -> {
                sendInvite()
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onResume() {
        super.onResume()

        hideKeyboard()
    }

    private fun hideKeyboard() {
        val activity = activity
        val imm = activity!!.getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
        var view = activity.currentFocus
        if (view == null) {
            view = View(activity)
        }
        imm.hideSoftInputFromWindow(view.windowToken, 0)
    }

    private fun sendInvite() {
        val sendIntent = Intent()
        val msg = getString(R.string.invite_content)
        sendIntent.action = Intent.ACTION_SEND
        sendIntent.putExtra(Intent.EXTRA_TEXT, msg + tasksListViewModel.deepLinkOfCurrentProject)
        sendIntent.type = "text/plain"
        startActivity(sendIntent)
    }

    private fun fillViews() {
        collectOnStarted(tasksListViewModel.projectReference) {
            setupActionBar(it?.projectName)
        }

        binding.addTask.setOnClickListener { v -> onNewTaskClick() }
        initRecyclerView()
    }

    private fun setupActionBar(projectName: String?) {
        val actionBar = (activity as AppCompatActivity).supportActionBar ?: return
        actionBar.title = projectName
    }

    private fun initRecyclerView() {
        collectOnStarted(tasksListViewModel.tasks) { tasks ->
            adapter.submitList(tasks)
        }
    }

    private fun onNewTaskClick() {
        navigationController.navigateToTaskDetail(null)
    }

    companion object {

        private val PROJECT_UUID_KEY = "projectUUID"

        fun create(projectUUID: String): TasksListFragment {
            val tasksListFragment = TasksListFragment()
            val args = Bundle()
            args.putString(PROJECT_UUID_KEY, projectUUID)
            tasksListFragment.arguments = args
            return tasksListFragment
        }
    }

}
